### PR TITLE
Fix exp10 fallback and exp2 and exp10 for Float16

### PR DESCRIPTION
**Pull request by musm**
`exp2` was  not added to the  Float16 definitions and hence calls the very slow fallback to `2^x`. 

`exp10` does not have any fallback definitions and thus things like 
```julia
julia> exp10(Float16(2.0))
ERROR: StackOverflowError:
```
lead to a StackOverFlow  or with other `AbstractFloat`

### CODE REVIEW

File: test/math.jl [hunk 1]
```diff
@@ -661,3 +661,17 @@ end
     @test sincos(big(1)) == (sin(big(1)), cos(big(1)))
     @test sincos(big(1.0)) == (sin(big(1.0)), cos(big(1.0)))
 end
+
+@testset "Float16 fallbacks" begin
```
Comment (fredrikekre) [inline test/math.jl]: Perhaps integrate with https://github.com/JuliaLang/julia/blob/e5cc4d7cdecf2eac83179564f4609e2b749321bf/test/math.jl#L252 ?

File: base/math.jl [hunk 2]
```diff
@@ -263,14 +263,27 @@ cbrt(x::AbstractFloat) = x < 0 ? -(-x)^(1//3) : x^(1//3)
 """
     exp2(x)
 
-Compute ``2^x``.
+Compute the base `2` exponential of `x`, in other words ``2^x``.
 
 ```jldoctest
```
Comment (fredrikekre) [inline base/math.jl]:
````
# Examples
```jldoctest
````

File: base/math.jl [hunk 3]
```diff
@@ -263,14 +263,27 @@ cbrt(x::AbstractFloat) = x < 0 ? -(-x)^(1//3) : x^(1//3)
...
 ```
 """
 exp2(x::AbstractFloat) = 2^x
+
+"""
+    exp10(x)
+
+Compute the base `10` exponential of `x`, in other words ``10^x``.
+
+```jldoctest
```
Comment (fredrikekre) [inline base/math.jl]:
````
# Examples
```jldoctest
````

Comment (musm) [inline base/math.jl]: thanks will fix later

File: test/math.jl (hunk 1 above)
Comment (musm) [inline test/math.jl]: done

File: test/math.jl [hunk 4]
```diff
@@ -661,3 +656,23 @@ end
...
+    @test exp10(50//10) ≈ exp10(5.0)
+    @test log10(exp10(e)) ≈ e
+    @test log(e) === 1
+    @test exp2(Float16(2.0)) ≈ exp2(2.0)
+    @test exp2(Float16(1.0)) === Float16(exp2(1.0))
+    @test exp10(Float16(1.0)) === Float16(exp10(1.0))
+end
+
+# test AbstractFloat fallback
+struct Float32432{T<:AbstractFloat} <: AbstractFloat
```
Comment (tkelman) [inline test/math.jl]: what is this number referring to?

Comment (musm) [inline test/math.jl]: Nothing, just a random string of numbers so it doesn't clash with another name

Comment (tkelman) [inline test/math.jl]: use the issue number or a word, less confusing that way

Comment (musm) [inline test/math.jl]: I change to use the PR number since I didn't open an issue for this

**Comment by musm** [comment]
Appveyor looks missing, restarting

**Comment by musm** [comment]
Appveyor looks missing, restarting

**Comment by musm** [comment]
Anything else to do here?

File: base/math.jl [hunk 5]
```diff
@@ -263,14 +263,29 @@ cbrt(x::AbstractFloat) = x < 0 ? -(-x)^(1//3) : x^(1//3)
 """
     exp2(x)
 
-Compute ``2^x``.
+Compute the base `2` exponential of `x`, in other words ``2^x``.
```
Comment (fredrikekre) [inline base/math.jl]: 2 should probably not be code highlighted? Likewise 10 in the docstring below.

Comment (musm) [inline base/math.jl]: I don't see a harm in highlighting the number, maybe others have  a different thought on this? Will update accordingly.

Comment (tkelman) [inline base/math.jl]: don't think it needs it for just the number on its own

**Pull request merged**